File: tricoach-backend/src/modules/plans/engine/microcycleBuilder.ts
import { randomUUID } from 'crypto';
import { LoadCalculator } from './loadCalculator';
import { addDays } from './periodizationEngine';
import { MacrocyclePhase, Mesocycle, Microcycle, ORDERED_WEEKDAYS, PlanGenerationContext } from './types';
import { WorkoutFactory } from './workoutFactory';

function rawWeekOrder(day: number): number {
  return ORDERED_WEEKDAYS.indexOf(day);
}

function weekdayDates(weekStart: Date): Map<number, Date> {
  const result = new Map<number, Date>();
  for (let offset = 0; offset < 7; offset++) {
    const date = addDays(weekStart, offset);
    result.set(date.getDay() + 1, date); // JS getDay(): Sun=0..Sat=6 -> Calendar-style Sun=1..Sat=7
  }
  return result;
}

/** Progressive overload within a 4-week block (weeks 1-3 build, week 4 unloads), tapering near the race. */
function loadMultiplier(globalWeekIndex: number, isRecoveryWeek: boolean, phase: MacrocyclePhase, totalPlanWeeks: number): number {
  if (phase === 'taper') {
    const weeksToGo = totalPlanWeeks - globalWeekIndex;
    return weeksToGo <= 0 ? 0.4 : 0.65;
  }
  if (isRecoveryWeek) return 0.65;
  const positionInBlock = (globalWeekIndex - 1) % 4;
  return Math.min(1.24, 1.0 + 0.08 * positionInBlock);
}

/** Direct port of Core/PlanEngine/MicrocycleBuilder.swift. */
export const MicrocycleBuilder = {
  build(mesocycle: Mesocycle, phase: MacrocyclePhase, startingGlobalWeekIndex: number, totalPlanWeeks: number, context: PlanGenerationContext): Microcycle[] {
    const totalDays = Math.round((mesocycle.endDate.getTime() - mesocycle.startDate.getTime()) / 86400000) + 1;
    const weekCount = Math.max(1, Math.round(totalDays / 7));

    const restDays = new Set(context.availability.mandatoryRestDays);
    const availableDays = context.availability.availableDays
      .filter((d) => !restDays.has(d))
      .sort((a, b) => rawWeekOrder(a) - rawWeekOrder(b));
    const sessionsPerWeek = Math.min(context.availability.sessionsPerWeek, Math.max(availableDays.length, 1));

    const microcycles: Microcycle[] = [];
    let weekStart = new Date(mesocycle.startDate);

    for (let offset = 0; offset < weekCount; offset++) {
      const globalIndex = startingGlobalWeekIndex + offset;
      const weekEnd = addDays(weekStart, 6);
      const isRecoveryWeek = phase !== 'taper' && phase !== 'transition' && globalIndex % 4 === 0;
      const multiplier = loadMultiplier(globalIndex, isRecoveryWeek, phase, totalPlanWeeks);

      const slots = WorkoutFactory.sessionSlots(sessionsPerWeek, availableDays, context.goal.type, phase, isRecoveryWeek, globalIndex);
      const datesByWeekday = weekdayDates(weekStart);

      const workouts = slots
        .map((slot) =>
          WorkoutFactory.makeWorkout(slot, datesByWeekday.get(slot.day) ?? weekStart, context.profile, context.availability.maxSessionDurationMin, multiplier)
        )
        .sort((a, b) => a.date.getTime() - b.date.getTime());

      microcycles.push({
        id: randomUUID(),
        weekNumber: globalIndex,
        startDate: weekStart,
        endDate: weekEnd,
        isRecoveryWeek,
        plannedLoad: LoadCalculator.totalLoad(workouts),
        workouts,
      });
      weekStart = addDays(weekStart, 7);
    }
    return microcycles;
  },
};

File: TriCoachAI/Core/PlanEngine/MicrocycleBuilder.swift
import Foundation

/// Builds the week-by-week microcycles for a mesocycle: assigns a load
/// multiplier per week (progressive overload with a lighter 4th week) and
/// delegates session placement to `WorkoutFactory`.
enum MicrocycleBuilder {
    static func build(
        mesocycle: Mesocycle,
        phase: MacrocyclePhase,
        startingGlobalWeekIndex: Int,
        totalPlanWeeks: Int,
        context: PlanGenerationContext
    ) -> [Microcycle] {
        let calendar = Calendar.current
        let totalDays = (calendar.dateComponents([.day], from: mesocycle.startDate, to: mesocycle.endDate).day ?? 6) + 1
        let weekCount = max(1, Int((Double(totalDays) / 7.0).rounded()))

        let restDays = context.availability.mandatoryRestDays
        let availableDays = context.availability.availableDays.subtracting(restDays)
            .sorted { $0.rawWeekOrder < $1.rawWeekOrder }
        let sessionsPerWeek = min(context.availability.sessionsPerWeek, max(availableDays.count, 1))

        var microcycles: [Microcycle] = []
        var weekStart = mesocycle.startDate

        for offset in 0..<weekCount {
            let globalIndex = startingGlobalWeekIndex + offset
            let weekEnd = calendar.date(byAdding: .day, value: 6, to: weekStart) ?? weekStart
            let isRecoveryWeek = phase != .taper && phase != .transition && globalIndex % 4 == 0
            let multiplier = loadMultiplier(globalWeekIndex: globalIndex, isRecoveryWeek: isRecoveryWeek, phase: phase, totalPlanWeeks: totalPlanWeeks)

            let slots = WorkoutFactory.sessionSlots(
                sessionsPerWeek: sessionsPerWeek,
                availableDays: availableDays,
                goalType: context.goal.type,
                phase: phase,
                isRecoveryWeek: isRecoveryWeek,
                weekIndexInPlan: globalIndex
            )

            let datesByWeekday = weekdayDates(from: weekStart, calendar: calendar)
            let workouts = slots.map { slot in
                WorkoutFactory.makeWorkout(
                    slot: slot,
                    date: datesByWeekday[slot.day] ?? weekStart,
                    profile: context.profile,
                    maxDurationMin: context.availability.maxSessionDurationMin,
                    loadMultiplier: multiplier
                )
            }.sorted { $0.date < $1.date }

            microcycles.append(
                Microcycle(
                    weekNumber: globalIndex,
                    startDate: weekStart,
                    endDate: weekEnd,
                    isRecoveryWeek: isRecoveryWeek,
                    plannedLoad: LoadCalculator.totalLoad(workouts),
                    workouts: workouts
                )
            )
            weekStart = calendar.date(byAdding: .day, value: 7, to: weekStart) ?? weekStart
        }
        return microcycles
    }

    /// Progressive overload within a 4-week block (weeks 1-3 build, week 4
    /// unloads), tapering down as the plan approaches its final race week.
    private static func loadMultiplier(globalWeekIndex: Int, isRecoveryWeek: Bool, phase: MacrocyclePhase, totalPlanWeeks: Int) -> Double {
        if phase == .taper {
            let weeksToGo = totalPlanWeeks - globalWeekIndex
            return weeksToGo <= 0 ? 0.40 : 0.65
        }
        if isRecoveryWeek {
            return 0.65
        }
        let positionInBlock = (globalWeekIndex - 1) % 4 // 0, 1, 2 are load weeks
        return min(1.24, 1.0 + 0.08 * Double(positionInBlock))
    }

    private static func weekdayDates(from weekStart: Date, calendar: Calendar) -> [Weekday: Date] {
        var result: [Weekday: Date] = [:]
        for dayOffset in 0..<7 {
            guard let date = calendar.date(byAdding: .day, value: dayOffset, to: weekStart) else { continue }
            let weekdayComponent = calendar.component(.weekday, from: date)
            if let weekday = Weekday(rawValue: weekdayComponent) {
                result[weekday] = date
            }
        }
        return result
    }
}

private extension Weekday {
    var rawWeekOrder: Int {
        Weekday.orderedWeek.firstIndex(of: self) ?? 0
    }
}
